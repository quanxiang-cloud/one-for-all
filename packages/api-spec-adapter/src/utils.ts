import type { Spec, Operation } from './swagger-schema-official';

import { METHODS } from './types';

export function indexOperation(spec: Spec): Record<string, Operation> {
  const operationMap: Record<string, Operation> = {};
  for (const [pathName, pathObj] of Object.entries(spec.paths || {})) {
    if (!pathObj) {
      continue;
    }

    for (const method of METHODS) {
      const operationObject = pathObj[method];
      if (!operationObject) {
        continue;
      }

      operationMap[`${method}:${pathName}`] = operationObject;
    }
  }

  return operationMap;
}

// copied from https://gist.github.com/creationix/7435851#gistcomment-3698888
// return an absolute path
export function join(...segments: string[]): string {
  const parts = segments.reduce<string[]>((parts, segment) => {
    // Remove leading slashes from non-first part.
    // Remove trailing slashes.
    const fragment = segment.replace(/^\//, '').replace(/\/$/, '');
    if (!fragment) {
      return parts;
    }

    const fragments = fragment.split('/');
    return parts.concat(fragments);
  }, []);

  const resultParts = [];
  for (const part of parts) {
    if (part === '.') {
      continue;
    }
    if (part === '..') {
      resultParts.pop();
      continue;
    }
    resultParts.push(part);
  }
  return `/${resultParts.join('/')}`;
}

type KeyPathPair = [string, string | number | boolean | null];

function isPrimaryValue(v: unknown): boolean {
  return typeof v === 'boolean' ||
    typeof v === 'number' ||
    typeof v === 'string' ||
    v === null;
}

function shouldGoIn(v: unknown): boolean {
  return typeof v === 'object' && v !== null;
}

function _toKeyPathPair(v: unknown, pairs: Array<KeyPathPair>, parentPath: string): Array<KeyPathPair> {
  if (typeof v !== 'object' || v === null) {
    return [];
  }

  Object.entries(v).forEach(([key, value]) => {
    const keyPath = parentPath ? `${parentPath}.${key}` : key;
    if (isPrimaryValue(value)) {
      pairs.push([keyPath, value]);
      return;
    }

    if (!shouldGoIn(v)) {
      return;
    }

    _toKeyPathPair((v as any)[key], pairs, keyPath);
  });

  return pairs;
}

export function toKeyPathPair(v: unknown, prefix: string): Array<KeyPathPair> {
  return _toKeyPathPair(v, [], prefix);
}
