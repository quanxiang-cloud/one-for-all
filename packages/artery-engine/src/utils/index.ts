export * from './build'
export * from './predicate';
