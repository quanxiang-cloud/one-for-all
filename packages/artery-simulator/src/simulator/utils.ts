import { parentIdsSeq } from '@one-for-all/artery-utils';
import { useEffect, useState } from 'react';
import { BehaviorSubject } from 'rxjs';
import { removeIn } from 'immutable';
import { generateNodeId } from '@one-for-all/artery-engine';
import type { Node } from '@one-for-all/artery';
import {
  byArbitrary,
  _appendTo,
  ImmutableNode,
  _insertLeftSiblingTo,
  _insertRightSiblingTo,
  ComposedNode,
  travel,
} from '@one-for-all/artery-utils';

import type { GreenZone, Cursor, Position, NodePrimary } from '../types';

const img = new Image();
img.src = 'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';

export function overrideDragImage(dateTransfer: DataTransfer): void {
  dateTransfer.setDragImage(img, 0, 0);
}

interface GetPositionParam {
  cursor: Cursor;
  hoveringRect: DOMRectReadOnly;
  supportInner: boolean;
}

// TODO optimize this
export function calcHoverPosition({ cursor, hoveringRect, supportInner }: GetPositionParam): Position {
  const leftDistance = Math.abs(cursor.x - hoveringRect.left);
  const rightDistance = Math.abs(cursor.x - hoveringRect.right);
  if (!supportInner) {
    return leftDistance < rightDistance ? 'left' : 'right';
  }

  if (leftDistance <= 9) {
    return 'left';
  }

  if (rightDistance <= 9) {
    return 'right';
  }

  const oneThirdWidth = hoveringRect.width / 3;
  if (leftDistance < oneThirdWidth) {
    return 'inner-left';
  }

  if (rightDistance < oneThirdWidth) {
    return 'inner-right';
  }

  return 'inner';
}

export function isChildNodeOf(
  root: Immutable.Collection<unknown, unknown>,
  parentID: string,
  childID: string,
): boolean {
  const parentIDs = parentIdsSeq(root, childID);
  if (!parentIDs) {
    return false;
  }

  return parentIDs.keyOf(parentID) !== undefined ? true : false;
}

export function getNodeExecutor(node: NodePrimary): string {
  if (node.type === 'html-element') {
    return `html-element:${node.name}`;
  }

  if (node.type === 'react-component') {
    return `react_component:${node.packageName}:${node.packageVersion}:${node.exportName}`;
  }

  return '';
}

export function useBehaviorSubjectState<T>(subject: BehaviorSubject<T>): T {
  const [state, setState] = useState(subject.value);

  useEffect(() => {
    const subscription = subject.subscribe(setState);
    return () => {
      subscription.unsubscribe();
    };
  }, [subject]);

  return state;
}

export function jsonParse<T>(json: string): T | undefined {
  try {
    return JSON.parse(json);
  } catch (error) {
    return;
  }
}

interface MoveNodeParams {
  rootNode: ImmutableNode;
  nodeID: string;
  greenZone: GreenZone;
}

export function moveNode({ rootNode, nodeID, greenZone }: MoveNodeParams): ImmutableNode | undefined {
  let _rootNode: ImmutableNode = rootNode;
  const nodeToMoveKeyPath = byArbitrary(rootNode, nodeID);
  if (!nodeToMoveKeyPath) {
    return;
  }

  const nodeToMove = rootNode.getIn(nodeToMoveKeyPath) as ImmutableNode | undefined;
  if (!nodeToMove) {
    return;
  }

  _rootNode = removeIn(rootNode, nodeToMoveKeyPath);

  return insertNode({ rootNode: _rootNode, node: nodeToMove, greenZone });
}

interface InsertNodeParams {
  rootNode: ImmutableNode;
  node: Node | ImmutableNode;
  greenZone: GreenZone;
}

export function insertNode({ rootNode, node, greenZone }: InsertNodeParams): ImmutableNode | undefined {
  if (
    greenZone.type === 'node_without_children' &&
    (greenZone.position === 'inner' ||
      greenZone.position === 'inner-left' ||
      greenZone.position === 'inner-right')
  ) {
    return _appendTo(rootNode, greenZone.contour.id, node);
  }

  if (greenZone.type === 'node_without_children' && greenZone.position === 'left') {
    return _insertLeftSiblingTo(rootNode, greenZone.contour.id, node);
  }

  if (greenZone.type === 'node_without_children' && greenZone.position === 'right') {
    return _insertRightSiblingTo(rootNode, greenZone.contour.id, node);
  }

  if (greenZone.type === 'adjacent-with-parent' && greenZone.edge === 'left') {
    return _insertLeftSiblingTo(rootNode, greenZone.child.id, node);
  }

  if (greenZone.type === 'adjacent-with-parent' && greenZone.edge === 'right') {
    return _insertRightSiblingTo(rootNode, greenZone.child.id, node);
  }

  if (greenZone.type !== 'between-nodes') {
    return;
  }

  if (greenZone.left.absolutePosition.height < greenZone.right.absolutePosition.height) {
    return _insertRightSiblingTo(rootNode, greenZone.left.id, node);
  }

  if (greenZone.left.absolutePosition.height > greenZone.right.absolutePosition.height) {
    return _insertLeftSiblingTo(rootNode, greenZone.right.id, node);
  }

  return _insertRightSiblingTo(rootNode, greenZone.left.id, node);
}

function regenerateNodeID<T extends Node | ComposedNode>(node: T): T {
  node.id = generateNodeId(node.type);

  return node;
}

// todo optimize performance
export function duplicateNode(node: Node): Node {
  const newNode = travel(node, {
    htmlNode: (current) => regenerateNodeID(current),
    reactComponentNode: (current) => regenerateNodeID(current),
    loopContainerNode: (current) => regenerateNodeID(current),
    composedNode: (current) => regenerateNodeID(current),
    refNode: (current) => regenerateNodeID(current),
    jsxNode: (current) => regenerateNodeID(current),
    routeNode: (current) => regenerateNodeID(current),
  });

  return newNode;
}
