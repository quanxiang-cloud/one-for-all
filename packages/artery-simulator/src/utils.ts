export function debounce<F extends (...args: any[]) => any>(
  fn: F,
): (...args: Parameters<F>) => Promise<ReturnType<F>> {
  let timeout: number;
  const debouncedFn = (...args: Parameters<F>): Promise<ReturnType<F>> => {
    cancelAnimationFrame(timeout);
    return new Promise<ReturnType<F>>((resolve) => {
      timeout = requestAnimationFrame(() => {
        const result: ReturnType<F> = fn(...args);
        resolve(result);
      });
    });
  };

  return debouncedFn;
}
