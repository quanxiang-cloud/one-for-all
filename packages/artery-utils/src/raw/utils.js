export function exists(value) {
  return value !== null && typeof value !== 'undefined';
}

export function getChildNodeKey(node) {
  const nodeType = node.getIn(['type']);
  if (nodeType === 'jsx-node' || nodeType === 'ref-node') {
    return;
  }

  if (nodeType === 'composed-node') {
    return 'nodes';
  }

  if (nodeType === 'loop-container' || nodeType === 'route-node') {
    return 'node';
  }

  return 'children';
}
