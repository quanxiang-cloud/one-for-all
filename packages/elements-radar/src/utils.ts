import type { Rect, Report } from './type';

const rectKeys: Array<keyof Rect> = ['height', 'width', 'x', 'y'];

export function isSame(previous: Report, current: Report): boolean {
  if (previous.size !== current.size) {
    return false;
  }

  return Array.from(previous.entries()).every(([element, { raw }]) => {
    const newRect = current.get(element);
    if (!newRect) {
      return false;
    }

    return rectKeys.every((key) => newRect.raw[key] === raw[key]);
  });
}

function calcRect(raw: DOMRectReadOnly, rootBounds: DOMRectReadOnly | null): Rect {
  const X = rootBounds?.x || 0;
  const Y = rootBounds?.y || 0;

  const rect: Rect = {
    height: Math.round(raw.height),
    width: Math.round(raw.width),
    x: Math.round(raw.x - X),
    y: Math.round(raw.y - Y),
  };

  return rect;
}

export function getReport(elements: HTMLElement[], root: HTMLElement): Promise<Report> {
  return new Promise((resolve) => {
    const observer = new IntersectionObserver(
      (entries) => {
        const pairs: Array<[HTMLElement, { raw: DOMRectReadOnly; rect: Rect }]> = entries.map(
          ({ target, boundingClientRect, rootBounds }) => {
            const rect: Rect = calcRect(boundingClientRect, rootBounds);
            return [target as HTMLElement, { rect, raw: boundingClientRect }];
          },
        );

        resolve(new Map(pairs));
      },
      { root },
    );

    elements.forEach((ele) => observer.observe(ele));
  });
}
