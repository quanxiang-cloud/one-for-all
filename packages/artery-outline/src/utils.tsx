import { useMemo } from 'react';
import { Node } from '@one-for-all/artery';
import {
  ancestors,
  ImmutableNode,
  KeyPath,
  _flat,
  _insertChildAt,
  _insertRightSiblingTo,
} from '@one-for-all/artery-utils';
import { fromJS, List } from 'immutable';
import { arrayMove } from '@dnd-kit/sortable';
import { logger } from '@one-for-all/utils';

import { INDENTATION_WIDTH } from './constants';
import { Entry, NodePrimary } from './types';

const NODE_TYPE_WHITE_LIST = ['html-element', 'react-component', 'loop-container'];

export function useFlattenNodes(
  rootNode: Node,
  collapsedNodeIDs: string[],
  draggingId?: string,
): Array<[KeyPath, ImmutableNode]> {
  const rawPairs = useMemo(() => {
    let pairs: Array<[KeyPath, ImmutableNode]> = _flat(fromJS(rootNode)).slice();
    for (let i = 0; i < pairs.length; ) {
      const nodeType = pairs[i][1].getIn(['type']);
      if (NODE_TYPE_WHITE_LIST.includes(nodeType as string)) {
        i = i + 1;
        continue;
      }

      const keyPathStr: string = pairs[i][0].toJS().join('/');
      while (pairs[i]?.[0].toJS().join('/').startsWith(keyPathStr)) {
        pairs.splice(i, 1);
      }
    }

    return pairs;
  }, [rootNode]);

  return useMemo(() => {
    let _rawPairs = rawPairs.slice();
    const idsShouldRemoveChildren = draggingId
      ? [draggingId, ...Array.from(collapsedNodeIDs)]
      : Array.from(collapsedNodeIDs);

    for (const id of idsShouldRemoveChildren) {
      const parentIndex = _rawPairs.findIndex(([_, n]) => n.getIn(['id']) === id);
      if (parentIndex === -1) {
        continue;
      }

      const [keyPath] = _rawPairs[parentIndex];
      const keyPathStr = keyPath.toJS().join('/');
      let lastDeleteIndex = parentIndex;
      for (let i = lastDeleteIndex + 1; i < _rawPairs.length; i += 1) {
        if (_rawPairs[i][0].toJS().join('/').startsWith(keyPathStr)) {
          lastDeleteIndex = i;
          continue;
        }

        break;
      }

      _rawPairs.splice(parentIndex + 1, lastDeleteIndex - parentIndex);
    }

    return _rawPairs;
  }, [draggingId, collapsedNodeIDs, rawPairs]);
}

export function pairToEntry(
  [keyPath, n]: [KeyPath, ImmutableNode],
  isContainer: (node: NodePrimary) => boolean,
): Entry {
  const id = n.getIn(['id']) as string;
  const primaryNode: NodePrimary = {
    type: n.getIn(['type']),
    name: n.getIn(['name']),
    packageName: n.getIn(['packageName']),
    packageVersion: n.getIn(['packageVersion']),
    exportName: n.getIn(['exportName']),
  } as NodePrimary;

  const _keyPath = keyPath.toArray().filter((path) => path === 'children' || path === 'nodes');

  return {
    id: id,
    name: (n.getIn(['label']) as string) || id,
    depth: _keyPath.length,
    isContainer: isContainer(primaryNode),
  };
}

export function getProjectionDepth(
  entries: Entry[],
  draggingId: string,
  overId: string,
  dragOffset: number,
): number {
  const overEntryIndex = entries.findIndex(({ id }) => id === overId);
  const draggingEntryIndex = entries.findIndex(({ id }) => id === draggingId);
  const draggingEntry = entries[draggingEntryIndex];
  const newEntries = arrayMove(entries, draggingEntryIndex, overEntryIndex);

  const previousEntry = newEntries[overEntryIndex - 1];
  const nextEntry = newEntries[overEntryIndex + 1];
  const dragDepth = Math.round(dragOffset / INDENTATION_WIDTH);
  const projectedDepth = draggingEntry.depth + dragDepth;

  const maxDepth = getMaxDepth(previousEntry);
  const minDepth = nextEntry ? nextEntry.depth : 0;

  let depth = projectedDepth;
  if (projectedDepth >= maxDepth) {
    depth = maxDepth;
  } else if (projectedDepth < minDepth) {
    depth = minDepth;
  }

  return depth;
}

function getMaxDepth(entry?: Entry): number {
  if (!entry) {
    return 0;
  }

  return entry.isContainer ? entry.depth + 1 : entry.depth;
}

export function insertBelowTo(
  rootNode: ImmutableNode,
  targetEntry: Entry,
  projectedDepth: number,
  node: ImmutableNode,
): ImmutableNode | undefined {
  if (projectedDepth === targetEntry.depth) {
    logger.debug(`[artery-outline] move [${node.getIn(['id'])}] to right of [${targetEntry.id}]`);

    return _insertRightSiblingTo(rootNode, targetEntry.id, node);
  }

  if (projectedDepth > targetEntry.depth) {
    logger.debug(
      `[artery-outline] move [${node.getIn(['id'])}] to the first position of [${targetEntry.id}]'s children`,
    );

    return _insertChildAt(rootNode, targetEntry.id, 0, node);
  }

  const parentIndex = targetEntry.depth - projectedDepth - 1;
  const parentList = ancestors(rootNode, targetEntry.id) || List();
  if (!parentList || !parentList.size) {
    logger.error(`[artery-outline] fatal error, can not find over node parents`);
    return;
  }

  const parentKeyPath = parentList.get(parentIndex);
  if (!parentKeyPath) {
    logger.error(`[artery-outline] fatal error, no ancestor at index ${parentIndex}`);
    return;
  }

  logger.debug(`[artery-outline] move [${node.getIn(['id'])}] to right of [${parentKeyPath.toJS()}]`);

  return _insertRightSiblingTo(rootNode, parentKeyPath, node);
}
