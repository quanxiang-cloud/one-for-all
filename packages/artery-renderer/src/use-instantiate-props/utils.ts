import { BehaviorSubject, combineLatestWith, map, of, skip, tap } from 'rxjs';
import { logger } from '@one-for-all/utils';
import type { ComputedDependency } from '@one-for-all/artery';

import { CTX, StateConvertor } from '../types';

interface ConvertResultParams {
  state: unknown;
  convertor?: StateConvertor;
  fallback: unknown;
  propName: string;
}

export function convertState({ state, convertor, fallback, propName }: ConvertResultParams): unknown {
  if (convertor && state !== undefined) {
    try {
      return convertor(state) ?? fallback;
    } catch (error) {
      logger.error(
        `an error occurred while calling state convertor for prop: "${propName}"`,
        'with the following state and convertor:',
        '\n',
        state,
        '\n',
        convertor.toString(),
        '\n',
        'So return fallback instead, fallback:',
        fallback,
        '\n',
        '\n',
        error,
      );
      return fallback;
    }
  }

  return state ?? fallback;
}

interface GetComputedState$Props {
  propName: string;
  deps: ComputedDependency[];
  convertor: StateConvertor;
  ctx: CTX;
  fallback: unknown;
}

export function getComputedState$({
  propName,
  deps,
  convertor,
  ctx,
  fallback,
}: GetComputedState$Props): BehaviorSubject<unknown> {
  let _fallback = fallback;
  const deps$ = deps.map<BehaviorSubject<unknown>>(({ type, depID }) => {
    if (type === 'api_state') {
      return ctx.statesHubAPI.getState$(depID) as BehaviorSubject<unknown>;
    }

    if (type === 'node_state') {
      return ctx.statesHubShared.getNodeState$(depID);
    }

    return ctx.statesHubShared.getState$(depID);
  });
  const initialDeps = deps$.reduce((acc: Record<string, unknown>, dep$, index) => {
    const key = deps[index].depID;
    acc[key] = dep$.value;
    
    return acc;
  }, {});
  const state$ = new BehaviorSubject(
    convertState({
      state: initialDeps,
      convertor: convertor,
      fallback: fallback,
      propName,
    }),
  );

  of(true)
    .pipe(
      combineLatestWith(deps$),
      map((_, ..._dep) => {
        return convertState({
          state: _dep,
          convertor: convertor,
          fallback: _fallback,
          propName,
        });
      }),
      skip(1),
      tap((state) => {
        _fallback = state;
      }),
    )
    .subscribe((state) => state$.next(state));

  return state$;
}
