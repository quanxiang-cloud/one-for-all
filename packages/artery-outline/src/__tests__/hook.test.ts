import { useFlattenNodes } from '../utils';
import { act, renderHook } from '@testing-library/react-hooks/pure';
import rootNode from './fixtures/node';
import { ImmutableNode, KeyPath } from '@one-for-all/artery-utils';

function toReadableString(pairs: Array<[KeyPath, ImmutableNode]>): string[] {
  return pairs.map(([keyPath, node]) => `${keyPath.toJS().join('/')}:${node.getIn(['id'])}`);
}

test('useFlattenNodes_should_return_all_nodes', () => {
  const { unmount, result } = renderHook(() => useFlattenNodes(rootNode, []));

  expect(toReadableString(result.current)).toMatchSnapshot();
  unmount();
});

describe('useFlattenNodes_should_filter_collapsed_node_children', () => {
  test('filter_first_level_nodes', () => {
    const { unmount, result } = renderHook(() => useFlattenNodes(rootNode, ['container']));

    expect(toReadableString(result.current)).toMatchSnapshot();
    unmount();
  });

  test('filter_children_of_todo-input-container', () => {
    const { unmount, result } = renderHook(() => useFlattenNodes(rootNode, ['todo-input-container']));

    expect(toReadableString(result.current)).toMatchSnapshot();
    unmount();
  });

  test('filter_children_of_todo-list-loop-composedNode', () => {
    const { unmount, result } = renderHook(() => useFlattenNodes(rootNode, ['todo-list-loop-composedNode']));

    expect(toReadableString(result.current)).toMatchSnapshot();
    unmount();
  });

  test('filter_children_of_active_node_todo-list-loop-composedNode', () => {
    const { unmount, result } = renderHook(() =>
      useFlattenNodes(rootNode, [], 'todo-list-loop-composedNode'),
    );

    expect(toReadableString(result.current)).toMatchSnapshot();
    unmount();
  });
});
