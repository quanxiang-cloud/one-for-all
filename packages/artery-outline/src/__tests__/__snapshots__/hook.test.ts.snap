// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFlattenNodes_should_filter_collapsed_node_children filter_children_of_active_node_todo-list-loop-composedNode 1`] = `
Array [
  ":container",
  "children/0:ref-schema",
  "children/1:todo-input-container",
  "children/1/children/0:todo-input",
  "children/2:todo-list-loop-composedNode",
  "children/3:footer",
  "children/3/children/0:todo-filter",
]
`;

exports[`useFlattenNodes_should_filter_collapsed_node_children filter_children_of_todo-input-container 1`] = `
Array [
  ":container",
  "children/0:ref-schema",
  "children/1:todo-input-container",
  "children/2:todo-list-loop-composedNode",
  "children/2/node:compose-node-container",
  "children/2/node/nodes/0:todo-toggle",
  "children/2/node/nodes/1:todo-title",
  "children/2/node/nodes/2:todo-delete",
  "children/3:footer",
  "children/3/children/0:todo-filter",
]
`;

exports[`useFlattenNodes_should_filter_collapsed_node_children filter_children_of_todo-list-loop-composedNode 1`] = `
Array [
  ":container",
  "children/0:ref-schema",
  "children/1:todo-input-container",
  "children/1/children/0:todo-input",
  "children/2:todo-list-loop-composedNode",
  "children/3:footer",
  "children/3/children/0:todo-filter",
]
`;

exports[`useFlattenNodes_should_filter_collapsed_node_children filter_first_level_nodes 1`] = `
Array [
  ":container",
]
`;

exports[`useFlattenNodes_should_return_all_nodes 1`] = `
Array [
  ":container",
  "children/0:ref-schema",
  "children/1:todo-input-container",
  "children/1/children/0:todo-input",
  "children/2:todo-list-loop-composedNode",
  "children/2/node:compose-node-container",
  "children/2/node/nodes/0:todo-toggle",
  "children/2/node/nodes/1:todo-title",
  "children/2/node/nodes/2:todo-delete",
  "children/3:footer",
  "children/3/children/0:todo-filter",
]
`;
